### PR TITLE
 Set mount type HostToContainer for mounts that include container storage root

### PR DESCRIPTION
**Pull request by harche**
#### What type of PR is this?

/kind bug

#### Which issue(s) this PR fixes:

Fixes # https://issues.redhat.com/browse/OCPBUGS-14351

#### Does this PR introduce a user-facing change?

```release-note
 Set mount type HostToContainer for mounts that include container storage root
```

### CODE REVIEW

File: server/container_create_linux.go [hunk 1]
```diff
@@ -969,6 +969,9 @@ func addOCIBindMounts(ctx context.Context, ctr ctrfactory.Container, mountLabel,
 		if m.HostPath == "/" && dest == "/" {
 			log.Warnf(ctx, "Configuration specifies mounting host root to the container root.  This is dangerous (especially with privileged containers) and should be avoided.")
 		}
+		if m.HostPath == "/var/lib/containers" {
```
Comment (haircommander) [inline server/container_create_linux.go:972]: we also should check /var/ and /var/lib (and also /lib)

Comment (haircommander) [reply server/container_create_linux.go:972]: and `/`

File: server/container_create_linux.go [hunk 2]
```diff
@@ -969,6 +969,9 @@ func addOCIBindMounts(ctx context.Context, ctr ctrfactory.Container, mountLabel,
...
+			m.Propagation = types.MountPropagation_PROPAGATION_HOST_TO_CONTAINER
```
Comment (haircommander) [inline server/container_create_linux.go:973]: can we log an info to let the user know what's happening? 

I also wonder if this should be configurable to start... wdyt @cri-o/cri-o-maintainers

Comment (saschagrunert) [reply server/container_create_linux.go]: Can we infer `/var/lib/containers` from an existing configuration?

File: server/container_create_linux.go [hunk 3]
```diff
@@ -969,6 +969,16 @@ func addOCIBindMounts(ctx context.Context, ctr ctrfactory.Container, mountLabel,
 		if m.HostPath == "/" && dest == "/" {
 			log.Warnf(ctx, "Configuration specifies mounting host root to the container root.  This is dangerous (especially with privileged containers) and should be avoided.")
 		}
+
+		hostPaths := []string{"/var", "/var/lib", "/var/lib/containers", "/lib", "/"}
+
+		for _, path := range hostPaths {
```
Comment (haircommander) [inline server/container_create_linux.go:975]: another nit: can we name this variable something more informative? `pathsContainingStorageRoot` or something?

File: server/container_create_linux.go (hunk 2 above)
Comment (haircommander) [reply server/container_create_linux.go]: I am inclined to agree with @saschagrunert here. I would not say the priority here is high enough to not have a complete solution.
using https://pkg.go.dev/path/filepath#Rel may be a possibility to calculate whether a path contains the storage root

Comment (haircommander) [reply server/container_create_linux.go]: alternatively, maybe
`strings.Contains(storage.Root, m.HostPath)` may give us what we want
https://go.dev/play/p/OU5jiFvElL3

**Comment by haircommander** [comment]
a couple more nits: a test would be sweet as well, and the release note could mention all of the other paths that are configured this way

File: server/utils.go [hunk 4]
```diff
@@ -223,3 +223,28 @@ func (s *Server) FilterDisallowedAnnotations(toFind, toFilter map[string]string,
 
 	return s.config.Workloads.FilterDisallowedAnnotations(allowed, toFilter)
 }
+
+// isSubDirectoryOf checks if the base path contains the target path.
```
Comment (haircommander) [inline server/utils.go:227]: I think the behavior is more `containsSubDirectory`

File: server/utils.go [hunk 5]
```diff
@@ -223,3 +223,28 @@ func (s *Server) FilterDisallowedAnnotations(toFind, toFilter map[string]string,
```
Comment (haircommander) [inline server/utils.go:223]: nit I'm personally a fan of keeping a function closest to the function that uses it, and reserving utils files for when the function is needed across multiple files. WDYT about moving this and the associated test?

File: server/utils.go (hunk 4 above)
Comment (haircommander) [reply server/utils.go:227]: nevermind I misread, the name is correct

File: server/container_create_linux.go [hunk 6]
```diff
@@ -914,7 +914,7 @@ func clearReadOnly(m *rspec.Mount) {
 	m.Options = append(m.Options, "rw")
 }
 
-func addOCIBindMounts(ctx context.Context, ctr ctrfactory.Container, mountLabel, bindMountPrefix string, absentMountSourcesToReject []string, maybeRelabel, skipRelabel, cgroup2RW bool) ([]oci.ContainerVolume, []rspec.Mount, error) {
+func addOCIBindMounts(ctx context.Context, ctr ctrfactory.Container, mountLabel, bindMountPrefix string, absentMountSourcesToReject []string, maybeRelabel, skipRelabel, cgroup2RW bool, storageRoot string) ([]oci.ContainerVolume, []rspec.Mount, error) {
```
Comment (haircommander) [inline server/container_create_linux.go:917]: one day I want this to all be refactored, it's really hard to read and figure out what happens to a mount

Comment (harche) [reply server/container_create_linux.go:917]: +1

File: server/utils.go (hunk 5 above)
Comment (harche) [reply server/utils.go]: Moved from utils.go to container_create_linux.go

File: server/container_create_linux.go [hunk 7]
```diff
@@ -969,6 +969,12 @@ func addOCIBindMounts(ctx context.Context, ctr ctrfactory.Container, mountLabel,
 		if m.HostPath == "/" && dest == "/" {
 			log.Warnf(ctx, "Configuration specifies mounting host root to the container root.  This is dangerous (especially with privileged containers) and should be avoided.")
 		}
+
+		if isSubDirectoryOf(storageRoot, m.HostPath) {
+			log.Infof(ctx, "Mount propogration for the host path %s will be set to HostToContainer", m.HostPath)
```
Comment (haircommander) [inline server/container_create_linux.go:974]: nit: can you include the rational "as it includes the container storage root"

File: server/container_create_linux.go [hunk 8]
```diff
@@ -969,6 +969,12 @@ func addOCIBindMounts(ctx context.Context, ctr ctrfactory.Container, mountLabel,
 		if m.HostPath == "/" && dest == "/" {
```
Comment (haircommander) [inline server/container_create_linux.go:969]: I think we can drop this now

**Comment by haircommander** [comment]
/approve

LGTM, @cri-o/cri-o-maintainers PTAL

**Review by saschagrunert — APPROVED** [review]

**Pull request merged**